### ISSUE TITLE
Recommended workflow

### ISSUE THREAD

**Issue by itissid** [bmuschko/gradle-docker-plugin#212]
This is more of a thought experiment for a use case I have. I use a docker base image to build a package (it requires some non trivial environment to build). Then I install the built package and start a service using the installed binaries. Finally it uses a return code as a way to send pass/fail signal back to the developer. What is the best workflow you propose for this use case in terms of gradle steps?
Given that repeatedly building and installing the package takes enough time to make development slow.

One thing I thought of was:

1. Use a Tag image with a fixed value for any given step.
2. If an image with a tag exists just use that image to spawn container in the next step make changes and commit the container again with a fixed but different tag from the original image.
3. Have a clean step to delete all the images and start from scratch.

Let me know what you think and perhaps suggest improvements/alternatives.

best

**Comment by cdancy** [comment]
@itissid hey and thanks for checking in. Because I don't know all the details surrounding what you're trying to do I can't say for certain if that is the best path forward. It sounds sane enough from the description you gave.

The one thing I can say is that committing over and over again, while it does work, has the limitation that you can only have a finite number of layers to your container. Check out this ISSUE for more details: https://github.com/docker/docker/issues/1171

**Comment by orzeh** [comment]
I would do this more or less like that:
````groovy
task buildNonTrivialWithDockerBaseImage(type: DockerBuildImage)
````
Tasks uses `Dockerfile` with `FROM` set to your base image and have volume that contains built package.

Next:
````groovy
task createContainer(type: DockerCreateContainer, dependsOn 'buildNonTrivialWithDockerBaseImage') {
    targetImageId { buildNonTrivialWithDockerBaseImage.getImageId() }
   //...
}

task startServiceUsingBuiltPackage(type: DockerStartContainer, dependsOn: 'createContainer')
````
This starts container which installs package and run service.
````groovy
task checkResult(type: DockerWaitContainer, dependsOn: 'startServiceUsingBuiltPackage') {
    targetContainerId {startServiceUsingBuiltPackage.getContainerId()}
    doLast{
        if(getExitCode() != 0) {
            println "Container failed with exit code \${getExitCode()}"
        } else {
            println "Container successful"
        }
    }
}

task myWrokflow(dependsOn: 'checkResult')
````

This is just a sample and does not work for sure ;-)

Take a look at [these functional tests](https://github.com/bmuschko/gradle-docker-plugin/blob/master/src/functTest/groovy/com/bmuschko/gradle/docker/DockerWaitContainerFunctionalTest.groovy) which has similar workflow.

**Comment by cdancy** [comment]
@itissid closing issue. Feel free to re-open if you're still having issues.

**Comment by itissid** [comment]
Hi thanks for your reply @orzeh. I did end up using most  of your code(thanks!). But, one thing that has arisen in the last few days is that there are some intermediate, time intensive steps after I start a container using my built image. The final result is a running service. 

0. I labelled tasks into groups. There were about a dozen of them.
1. To speed up development of each task, I decided to run tasks only when tasks were "complete". Which typically was based on  the presence of a file:
```
gradle.taskGraph.beforetasks {
  Task t -> 
     t.only if stepIsComplete(t.getGroup())
}
```
I was not able to leverage the [inputs, outputs](https://docs.gradle.org/current/userguide/more_about_tasks.html#sec:up_to_date_checks) of tasks because input properties like imageid, containerid(among others) made it a chicken and problem and also brittle.
